src/common/tests/checksum_tests.c

```diff
@@ -1042,36 +1042,119 @@ test_is_csum_enabled(void **state)
+	fake_algo.cf_get_size = fake_get_size;
 
-	for (type = CSUM_TYPE_UNKNOWN + 1; type < CSUM_TYPE_END; type++) {
-		rc = daos_csummer_init(&csummer,
-				       daos_csum_type2algo(type), 128);
-		assert_int_equal(0, rc);
+	dts_sgl_init_with_strings(&sgl, 1, "abcdef");
 
-		iod.iod_nr = 1;
-		iod.iod_recxs = NULL;
-		iod.iod_size = daos_sgl_buf_size(&sgl);
-		iod.iod_type = DAOS_IOD_SINGLE;
+	iod.iod_nr = 1;
+	iod.iod_recxs = NULL;
+	iod.iod_size = 1; /* [todo-ryon]: this should be bigger than 1 */
+	iod.iod_type = DAOS_IOD_SINGLE;
 
-		rc = daos_csummer_calc_iods(csummer, &sgl, &iod, 1, &csums);
+	rc = daos_csummer_calc_iods(csummer, &sgl, &iod, 1, &actual);
 
-		assert_int_equal(0, rc);
+	assert_int_equal(0, rc);
 
-		daos_csummer_free_ic(csummer, &csums);
-		daos_csummer_destroy(&csummer);
-	}
+	assert_int_equal(fake_get_size_result, actual->ic_data->cs_buf_len);
+	assert_int_equal(1, actual->ic_data->cs_nr);
+	assert_int_equal(fake_get_size_result, actual->ic_data->cs_len);
+
+	assert_int_equal(1, *ci_idx2csum(actual->ic_data, 0));
 
+	daos_csummer_free_ic(csummer, &actual);
 	daos_sgl_fini(&sgl, true);
+	daos_csummer_destroy(&csummer);
+}
+
+static void
+test_compare_sv_checksums(void **state)
+{
+	struct daos_csummer	*csummer;
+	struct dcs_iod_csums	*one;
+	struct dcs_iod_csums	*two;
+	d_sg_list_t		 sgl;
+	daos_iod_t		 iod = {0};
+	int			 rc = 0;
+
+	fake_get_size_result = 4;
+
+	daos_csummer_init(&csummer, &fake_algo, 4);
+	dts_sgl_init_with_strings(&sgl, 1, "0123456789");
+
+	fake_update_bytes_seen = 0;
+
+	iod.iod_nr = 1;
+	iod.iod_recxs = NULL;
+	iod.iod_size = daos_sgl_buf_size(&sgl);
+	iod.iod_type = DAOS_IOD_SINGLE;
+
+	rc = daos_csummer_calc_iods(csummer, &sgl, &iod, 1, &one);
+	assert_int_equal(0, rc);
+	rc = daos_csummer_calc_iods(csummer, &sgl, &iod, 1, &two);
+	assert_int_equal(0, rc);
+
+	assert_true(daos_csummer_compare_csum_info(csummer, one->ic_data,
+						   two->ic_data));
+
+	daos_sgl_fini(&sgl, true);
+	daos_csummer_free_ic(csummer, &one);
+	daos_csummer_free_ic(csummer, &two);
+	daos_csummer_destroy(&csummer);
+}
+
+static void
+test_verify_sv_data(void **state)
+{
+	struct daos_csummer	*csummer;
+	daos_iod_t		 iod = {0};
+	d_sg_list_t		 sgl = {0};
+	daos_size_t		 sgl_buf_half;
+	int			 rc;
+	struct dcs_iod_csums	*iod_csums = NULL;
+
+	daos_csummer_type_init(&csummer, CSUM_TYPE_ISAL_CRC64_REFL, 1024*1024);
+	dts_sgl_init_with_strings(&sgl, 1, "0123456789");
+
+
+	iod.iod_size = daos_sgl_buf_size(&sgl);
+	iod.iod_nr = 1;
+	iod.iod_recxs = NULL;
+	iod.iod_type = DAOS_IOD_SINGLE;
+
+	/** Checksum not set in iod_csums but csummer is set so should error */
+	rc = daos_csummer_verify(csummer, &iod, &sgl, iod_csums);
+	assert_int_equal(-DER_INVAL, rc);
+
+	rc = daos_csummer_calc_iods(csummer, &sgl, &iod, 1, &iod_csums);
+	assert_int_equal(0, rc);
+
+	rc = daos_csummer_verify(csummer, &iod, &sgl, iod_csums);
+	assert_int_equal(0, rc);
+
+	((char *)sgl.sg_iovs[0].iov_buf)[0]++; /** Corrupt the data */
```

Comment:
This seems to be a very low level data corruption. Is it possible to use any of the existing checksum fault injection methods (or) enhance existing FIs. It will be useful for SYSTEM level validation in future.

#define DAOS_CHECKSUM_UPDATE_FAIL	(DAOS_FAIL_UNIT_TEST_GROUP_LOC | 0x20)
#define DAOS_CHECKSUM_FETCH_FAIL	(DAOS_FAIL_UNIT_TEST_GROUP_LOC | 0x21)
#define DAOS_CHECKSUM_CDATA_CORRUPT	(DAOS_FAIL_UNIT_TEST_GROUP_LOC | 0x22)
#define DAOS_CHECKSUM_SDATA_CORRUPT	(DAOS_FAIL_UNIT_TEST_GROUP_LOC | 0x23)

Reply:
I had most of this coded up before the existing checksum fault injection methods existed. Was just waiting on the checksum removal from the API. I'm not sure I want to go through CI again for fault injection changes :-). Is it alright if I make this change in a future patch?

Reply:
Absolutely... You can do it future patch. Agree, getting CI passing all tests is like winning lottery :-).